src/liballoc_frame/lib.rs

```diff
@@ -0,0 +1,191 @@
+// Copyright 2015 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#![crate_name = "alloc_frame"]
+#![crate_type = "rlib"]
+#![no_std]
+#![allocator]
+#![cfg_attr(not(stage0), deny(warnings))]
+#![unstable(feature = "alloc_frame",
+            reason = "this library is unlikely to be stabilized in its current \
+                      form or name",
+            issue = "0")]
+#![feature(allocator)]
+#![feature(const_fn)]
+#![feature(staged_api)]
+#![feature(libc)]
+
+extern crate libc;
+
+use core::ptr;
+
+// The minimum alignment guaranteed by the architecture. This value is used to
+// add fast paths for low alignment values. In practice, the alignment is a
+// constant at the call site and the branch will be optimized out.
+#[cfg(all(any(target_arch = "x86",
+              target_arch = "arm",
+              target_arch = "mips",
+              target_arch = "powerpc",
+              target_arch = "powerpc64",
+              target_arch = "asmjs",
+              target_arch = "wasm32")))]
+const MIN_ALIGN: usize = 8;
+#[cfg(all(any(target_arch = "x86_64",
+              target_arch = "aarch64",
+              target_arch = "mips64",
+              target_arch = "s390x")))]
+const MIN_ALIGN: usize = 16;
+
+// The size of the chunk which is actually allocated
+const CHUNK_SIZE: usize = 4096 * 16;
+const CHUNK_ALIGN: usize = 4096;
+
+static mut HEAP: *mut u8 = ptr::null_mut();
+static mut HEAP_LEFT: usize = 0;
```

Comment:
Oh, yep. I had implemented this using atomics the first time I wanted to do this (but got stuck getting the allocator to inject). I'll go dig that code up.